### PR TITLE
FIX: Make category-drop work with lazy_load_categories

### PR DESCRIPTION
**Pull request by nbianca**
The category drop was rerendered after every category async search because it updated the categories list stored in the Site instance. This is not necessary and categories can be referenced indirectly by ID instead to avoid loading objects.

This PR contains other various bug fixes to make it mostly work as expected.

### CODE REVIEW

File: app/controllers/categories_controller.rb [hunk 1]
```diff
@@ -299,6 +300,24 @@ def visible_groups
     render json: success_json.merge(groups: groups || [])
   end
 
+  def find
+    category = Category.find_by_slug_path_with_id(params[:category_slug_path_with_id])
+    raise Discourse::NotFound if !category.present?
```
Comment (tgxworld) [inline app/controllers/categories_controller.rb:305]:
```suggestion
    raise Discourse::NotFound if category.blank?
```

File: app/controllers/categories_controller.rb [hunk 2]
```diff
@@ -299,6 +300,24 @@ def visible_groups
     render json: success_json.merge(groups: groups || [])
   end
 
+  def find
```
Comment (tgxworld) [inline app/controllers/categories_controller.rb:303]: It looks like we might have missed out on a test for this controller action?

File: app/assets/javascripts/discourse/app/components/bread-crumbs.hbs [hunk 3]
```diff
@@ -17,17 +17,7 @@
...
-          (concat
-            "--category-bg-color:#"
-            breadcrumb.category.color
-            "; "
-            "--category-text-color:#"
-            breadcrumb.category.text_color
-            ";"
-          )
-        )
+        (if breadcrumb.category (category-variables breadcrumb.category))
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/bread-crumbs.hbs:20]: Just want to confirm that this is a refactoring change and not really related to the fix here?

File: app/assets/javascripts/discourse/app/components/bread-crumbs.js [hunk 4]
```diff
@@ -44,7 +31,7 @@ export default Component.extend({
         options,
         isSubcategory: !!parentCategory,
         noSubcategories: !category && noSubcategories,
-        hasOptions: options.length !== 0,
+        hasOptions: !parentCategory || parentCategory.has_children,
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/bread-crumbs.js:34]: I'm not sure I understand the need for`parentCategory.has_children` here because the fact that the `parentCategory` exists mean it has child categories. Otherwise, `parentCategory` would be blank.

File: app/assets/javascripts/discourse/app/components/bread-crumbs.js [hunk 5]
```diff
@@ -10,21 +10,8 @@ export default Component.extend({
...
-        category.id !== this.site.uncategorized_category_id
-    );
-  },
-
-  @discourseComputed(
-    "category.ancestors",
-    "filteredCategories",
-    "noSubcategories"
-  )
-  categoryBreadcrumbs(categoryAncestors, filteredCategories, noSubcategories) {
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/bread-crumbs.js:27]: I am not sure I follow the change here, why do we need to remove `filteredCategories` here?

File: app/assets/javascripts/discourse/app/instance-initializers/category-badge-css-generator.js [hunk 6]
```diff
@@ -17,9 +17,9 @@ export default {
       let parentCategory = Category.findById(
         get(category, "parent_category_id")
       );
-      let badgeCss = `.badge-category[data-category-id="${category.id}"] { --category-badge-color: var(--category-${category.id}-color); --category-badge-text-color: #${category.text_color}; }`;
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/instance-initializers/category-badge-css-generator.js:20]: I am wondering if this change is required as part of the fix. If not, I think we should split it out as a UX change in another PR>

File: app/assets/javascripts/discourse/app/routes/build-category-route.js [hunk 7]
```diff
@@ -29,9 +30,11 @@ class AbstractCategoryRoute extends DiscourseRoute {
   controllerName = "discovery/list";
 
   async model(params, transition) {
-    const category = Category.findBySlugPathWithID(
-      params.category_slug_path_with_id
-    );
+    const category = this.siteSettings.lazy_load_categories
+      ? await Category.asyncFindBySlugPathWithID(
+          params.category_slug_path_with_id
+        )
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/routes/build-category-route.js:36]: In the context of this PR, what problem is this solving? Also, I wonder if we should first check if the category has already been preloaded instead of triggering a network request to fetch the categories unconditionally.

File: app/assets/javascripts/select-kit/addon/components/category-drop.js [hunk 8]
```diff
@@ -54,6 +54,37 @@ export default ComboBoxComponent.extend({
     return this.options.subCategory || false;
   }),
 
+  shortcuts: computed(
```
Comment (tgxworld) [inline app/assets/javascripts/select-kit/addon/components/category-drop.js:57]: Will you be able to expand on why this change is required and what problem does this change solve? :bow:

File: app/models/category.rb [hunk 9]
```diff
@@ -302,8 +318,24 @@ def clear_subcategory_ids
...
+  def has_children?
+    !!id && Category.has_children?(id)
+  end
+
+  def self.clear_parent_ids
+    @@has_children.clear
+  end
+
+  def clear_parent_ids
+    Category.clear_parent_ids
```
Comment (tgxworld) [inline app/models/category.rb:338]: If we are calling `category.clear_parent_ids`, are we able to only clear the affected category instead of clearing the entire cache?

File: app/models/concerns/category_hashtag.rb [hunk 10]
```diff
@@ -40,7 +40,7 @@ def query_loaded_from_slugs(category_slugs, categories)
               end
             end
           else
-            categories.find { |cat| cat.slug.casecmp?(parent_slug) && cat.top_level? }
+            categories.find { |cat| cat.parent_category_id.nil? && cat.slug.casecmp?(parent_slug) }
```
Comment (tgxworld) [inline app/models/concerns/category_hashtag.rb:43]: Will you be able to expand on why this is being changed in the context of this PR?

File: app/models/site.rb [hunk 11]
```diff
@@ -5,7 +5,7 @@ class Site
   include ActiveModel::Serialization
 
   # Number of categories preloaded when lazy_load_categories is enabled
-  LAZY_LOAD_CATEGORIES_LIMIT = 50
+  LAZY_LOAD_CATEGORIES_LIMIT = 10
```
Comment (tgxworld) [inline app/models/site.rb:8]: Is there a reason we are reducing the limit here?

File: app/models/topic_list.rb [hunk 12]
```diff
@@ -78,7 +78,7 @@ def topics
 
   def categories
     @categories ||=
-      topics.map { |t| [t.category, t.category&.parent_category] }.uniq.flatten.compact
+      topics.map { |t| [t.category&.parent_category, t.category] }.uniq.flatten.compact
```
Comment (tgxworld) [inline app/models/topic_list.rb:81]: Curious why the order has to be changed here.

File: app/serializers/basic_category_serializer.rb [hunk 13]
```diff
@@ -19,7 +19,7 @@ class BasicCategorySerializer < ApplicationSerializer
              :notification_level,
              :can_edit,
              :topic_template,
-             :has_children,
+             :has_children?,
```
Comment (tgxworld) [inline app/serializers/basic_category_serializer.rb:22]: Is `?` allowed in the attribute name? Im wondering if there is an existing test for this.

File: app/controllers/categories_controller.rb [hunk 14]
```diff
@@ -334,10 +353,7 @@ def search
         ) if term.present?
 
     categories =
-      categories.where(
-        "id = :id OR parent_category_id = :id",
-        id: parent_category_id,
-      ) if parent_category_id.present?
```
Comment (tgxworld) [inline app/controllers/categories_controller.rb:340]: What is the motivation for changing this?

File: app/assets/javascripts/discourse/app/components/bread-crumbs.hbs (hunk 3 above)
Comment (nbianca) [reply app/assets/javascripts/discourse/app/components/bread-crumbs.hbs:20]: Refactoring and a bug fix. The category colors were not rendered correctly, because we are now injecting stylesheets to color the badges in an initializer, but also because the variables names are not consistent.

File: app/assets/javascripts/discourse/app/components/bread-crumbs.js (hunk 5 above)
Comment (nbianca) [reply app/assets/javascripts/discourse/app/components/bread-crumbs.js:27]: We used to filter categories here and in d-navigation, the component that renders this one. I moved all the filtering logic there, to keep dependencies between components simpler.

File: app/assets/javascripts/discourse/app/instance-initializers/category-badge-css-generator.js (hunk 6 above)
Comment (nbianca) [reply app/assets/javascripts/discourse/app/instance-initializers/category-badge-css-generator.js:20]: Hmm, I think I can move some of these to a different PR.

File: app/assets/javascripts/discourse/app/routes/build-category-route.js (hunk 7 above)
Comment (nbianca) [reply app/assets/javascripts/discourse/app/routes/build-category-route.js:36]: Yes and no. I mean it is not a direct fix to category-drop, but navigating using category-drop is broken otherwise.

File: app/assets/javascripts/select-kit/addon/components/category-drop.js (hunk 8 above)
Comment (nbianca) [reply app/assets/javascripts/select-kit/addon/components/category-drop.js:57]: There are a few branches like `if (this.siteSettings.lazy_load_categories) { ... } else { ... }` and this refactoring made me duplicate less code between the two branches.

File: app/models/category.rb (hunk 9 above)
Comment (nbianca) [reply app/models/category.rb:338]: I think we could clear only the old and new parent IDs.

File: app/models/concerns/category_hashtag.rb (hunk 10 above)
Comment (nbianca) [reply app/models/concerns/category_hashtag.rb:43]: This was just a small refactoring, I can move it to a different PR.

File: app/models/site.rb (hunk 11 above)
Comment (nbianca) [reply app/models/site.rb:8]: To make other category scaling bugs more obvious, but I can move to a different PR.

File: app/models/topic_list.rb (hunk 12 above)
Comment (nbianca) [reply app/models/topic_list.rb:81]: The order categories are sorted matters for the frotend to ensure that the `parentCategory` properties are set correctly.

File: app/serializers/basic_category_serializer.rb (hunk 13 above)
Comment (nbianca) [reply app/serializers/basic_category_serializer.rb:22]: There was a test, but it was broken. The serializer types were "string/null", but it should've been "bool".

Attribute names containing characters such as a question mark are sanitized and only `has_children` is returned. The question mark is necessary because we want to serialize `category.has_children?` (which exists) instead of `category.has_children` (which does not exist).

File: app/controllers/categories_controller.rb (hunk 14 above)
Comment (nbianca) [reply app/controllers/categories_controller.rb:340]: This was a bug. It returned the parent category too, when it should've returned only children categories.

File: app/assets/javascripts/discourse/app/components/bread-crumbs.hbs (hunk 3 above)
Comment (nbianca) [reply app/assets/javascripts/discourse/app/components/bread-crumbs.hbs:20]: Created https://github.com/discourse/discourse/pull/24527

File: app/controllers/categories_controller.rb (hunk 2 above)
Comment (nbianca) [reply app/controllers/categories_controller.rb:303]: I added a test.

File: app/models/topic_list.rb (hunk 12 above)
Comment (nbianca) [reply app/models/topic_list.rb]: I fixed that in another PR: https://github.com/discourse/discourse/pull/24530

File: app/controllers/categories_controller.rb (hunk 14 above)
Comment (tgxworld) [reply app/controllers/categories_controller.rb:340]: :+1:

**Review by tgxworld — APPROVED** [review]
